### DISCUSSION TITLE
System.Exception: "IBrowserHost instance is null. Browser has likely not finished initializing or is in the process of disposing."

### DISCUSSION THREAD

**Discussion by Includephone** [cefsharp/CefSharp#4031] [General]
Hello, i have a problem with adding cookies in Cookie Mananger. When i try to set cookie, i get System.Exception: "IBrowserHost instance is null. Browser has likely not finished initializing or is in the process of disposing."
The script works without errors when using Cef.GetGlobalCookieManager(), but I need to set the cookie to this particular browser instance

```c#
public partial class Form2 : Form
    {
        public Form2(Account account)
        {
            InitializeComponent();
            var browser = new ChromiumWebBrowser("www.google.com");
            RequestContextSettings requestContextSettings = new RequestContextSettings();
            requestContextSettings.PersistSessionCookies = false;
            requestContextSettings.PersistUserPreferences = false;
            requestContextSettings.CachePath = Path.Combine(Environment.GetFolderPath(Environment.SpecialFolder.LocalApplicationData), "CefSharp\\Cache\\" + account.steamId);
            browser.RequestContext = new RequestContext(requestContextSettings);
            foreach (System.Net.Cookie cookieLine in account.cookie)
            {
                if (cookieLine.Name == "steamLoginSecure")
                {
                    var cookieMananger = browser.GetCookieManager();
                    Thread.Sleep(5000);
                    cookieMananger.SetCookie("https://steamcommunity.com/", new CefSharp.Cookie
                    {
                        Name = cookieLine.Name,
                        Value = cookieLine.Value,
                        Expires = cookieLine.Expires
                    });
                }
            }
            browser.Load("https://steamcommunity.com/");
            this.Controls.Add(browser);
        }
    }
```

**Comment by amaitland** [comment]
In future please use discussions to ask questions. This is not a bug.

https://github.com/cefsharp/CefSharp/discussions

> Browser has likely not finished initializing

In your case the browser hasn't been initialized yet.

Wait for the browser to initialize before attempting to access the cookie manager. 

https://github.com/cefsharp/CefSharp.MinimalExample/blob/master/CefSharp.MinimalExample.WinForms/BrowserForm.cs#L33

**Comment by mwpowellhtx** [comment]
Getting similar results. I second this issue nomination. Maybe we are doing something funny discussion worthy, but so far, looking like a bug. Happy to be incorrect in that. CEF is definitely claiming Initialized, however, the BrowserCore is definitely when that is raised, `null`.

As far as usage, I am starting from the minimal WPF example use case. So far I am creating the window with the CEF instance embedded, however, I am leaving it `Visibility = Hidden` for the time being, unless/until I need it. Maybe because it is hidden, i.e. not shown, some GC is happening that _"should not"_, best guess. Behaviors are also involved almost verbatim from the minimal WPF samples, but I should not think those have any impact here, i.e. the _hoverlink behaviors_.

```xaml
<Border BorderBrush="Gray" BorderThickness="0,1">
    <wpf:ChromiumWebBrowser x:Name="Browser" Address="..." Initialized="Wrapper_Browser_Initialized" />
</Border>
```

```csharp
private void Wrapper_Browser_Initialized(object sender, EventArgs e)
{
    // sender is ChromiumWebBrowser, which BrowserCore is definitely null at the moment this is raised.
}
```

Under debugging watch:

![image](https://github.com/cefsharp/CefSharp/assets/1824187/7b2e2bfa-6f22-49f3-a3aa-426ca7f202e9)

**Reply by mwpowellhtx** [reply]
When the `Window` is otherwise visible, I have been able to verify that the browser `Address` is loading as expected, so at least at that level, I know things have some possibility of working. Then I shifted gears into `Hidden Visibility`, which as I said, maybe some GC liberties are being assumed there 🤷‍♂️ .

Plus the fact that the `Initialized` event is indeed being raised, at least as far as the CEF scaffold is concerned, we think it is initialized.

**Reply by mwpowellhtx** [reply]
Trying a couple of things, still somewhat in learning phase, but things are becoming a bit clearer. It is _very_ asynchronous, that much I can say. Doing a ton of async awaits, but they are at least evaluating, to a point, so that's a start, better than before.

The good news is, I can toggle visibility, with a subsequent await, so beyond that hurdle.

```csharp
Wrapper.Visibility = Visibility.Visible;

// After turning the Window visible, must wait for rendering to complete, but it does (!) render.
await Wrapper.Browser.WaitForRenderIdleAsync();

// Selector waits are working it seems like, or the functional loading has not fully rendered...
await Wrapper.Browser.WaitForSelectorAsync("input[type = 'text', name = 'username']");
await Wrapper.Browser.WaitForSelectorAsync("input[type = 'password', name = 'password']");
await Wrapper.Browser.WaitForSelectorAsync("input[type = 'checkbox', name = 'rememberme']");
await Wrapper.Browser.WaitForSelectorAsync("button[type = 'submit', name = 'submitLogin']");

await using var context = await Wrapper.Browser.CreateDevToolsContextAsync();

var ctrlUsername = await context.QuerySelectorAsync<HtmlInputElement>("input[type = 'text', name = 'username']");
var ctrlPassword = await context.QuerySelectorAsync<HtmlInputElement>("input[type = 'password', name = 'password']");
// TODO: Feed me with View Model username and password.

var cbRememberMe = await context.QuerySelectorAsync<HtmlInputElement>("input[type = 'checkbox', name = 'rememberme']");
// Assuming from within the scope of the View Model.
await cbRememberMe.SetCheckedAsync(RememberMe);

// After having fed the controls, click.
await context.ClickAsync("button[type = 'submit', name = 'submitLogin']");
```

By comparison and/or contrast, from inside a Chrome dev tools console, I do have controls where I think I should have controls using the same selectors, everything else being equal.

![image](https://github.com/cefsharp/CefSharp/assets/1824187/be169f71-23bd-4d58-9fa9-b18547a6f56c)

**Reply by mwpowellhtx** [reply]
Getting an exception that query selector on the document is invalid on line:

```csharp
var ctrlUsername = await context.QuerySelectorAsync<HtmlInputElement>("input[type = 'text', name = 'username']");
```

For whatever reason. It seems as though, either:

1. the `Wrapper.Browser.WaitForSelectorAsync(...)` is way too aggressive, not sure what it is watching for, the thing has not fully loaded; or
1. the subsequent `context.QuerySelectorAsync<HtmlInputElement>("...")` is unable to determine that the thing is the thing we thought it was, in this case, `<input/>`, for example; or
1. I cannot think of a third thing, outside the possibility that we need to properly align the selector with the generic, i.e. `HtmlInputElement` for `<input/>`, and so on.

Kind of at a loss, unless someone can offer some guidance, things to look for, getting this _DOM puppeteering session_ on its feet.

**Reply by mwpowellhtx** [reply]
Did a minor adjustment still no luck. Here is the exception text:

```csharp
await context.WaitForSelectorAsync("input[type='text', name='username']", new WaitForSelectorOptions { Timeout = 30000, Visible = true });
```

It fails immediately. Tried tick marks and full quotes in the selector string, neither case works. As I stated above, the same selector strings _"just work"_ for purposes of the Chrome dev tools console. So not sure why they would not work here.

```
CefSharp.Dom.EvaluationFailedException: 'Evaluation failed: DOMException: Failed to execute 'querySelector' on 'Document': 'input[type="text", name="username"]' is not a valid selector.
    at predicate (eval at waitForPredicatePageFunction (:3:23), <anonymous>:7:30)
    at eval (eval at waitForPredicatePageFunction (:3:23), <anonymous>:23:17)
    at onRaf (__puppeteer_evaluation_script__:50:35)
    at pollRaf (__puppeteer_evaluation_script__:43:15)
    at waitForPredicatePageFunction (__puppeteer_evaluation_script__:8:22)'
```

**Reply by mwpowellhtx** [reply]
Next bit of a clue; the selector should work under Chrome console itself. If it is not working there, then it will not work through CEF. Troubleshooting that now.

**Reply by amaitland** [reply]
> Failed to execute 'querySelector' on 'Document': 'input[type="text", name="username"]' is not a valid selector

That exception comes straight out of `V8`. 

> the selector should work under Chrome console itself. If it is not working there, then it will not work through CEF. Troubleshooting that now.

Use `DevTools` for debugging directly in the `ChromiumWebBrowser` instance. https://github.com/cefsharp/CefSharp/wiki/General-Usage#devtools

**Comment by mwpowellhtx** [comment]
For Pete's sake, VICTORY! 🎆 🔥

**Reply by mwpowellhtx** [reply]
That was it, the steps above, were more or less on point. And with a bit of DOM and query selector refinement. Got 'er working.

**Comment by mwpowellhtx** [comment]
For the issue itself, ensuring that the host Window is visible and has loaded fully, is the key for the Browser instance to also have been initialized and DOM available. Once that was the case, sorted out my Chrome selector query strings, which were also problematic. Once I got those sorted out, the rest flowed like magic seamlessly with the browser window.

**Comment by amaitland** [comment]
> Wait for the browser to initialize before attempting to access the cookie manager.
> 
> https://github.com/cefsharp/CefSharp.MinimalExample/blob/master/CefSharp.MinimalExample.WinForms/BrowserForm.cs#L33

The **WinForms** and **WPF** implementations both have an `IsBrowserInitializedChanged` event that's fired when the browser is initialzied. (The link above is no off by one line, time goes on).

> Getting similar results. I second this issue nomination. Maybe we are doing something funny discussion worthy, but so far, looking like a bug.

The code provided by the `OP` is incorrect (**not a bug**). Creating a browser instance then immediately attempting to access the cookie manager is never going to work. The same goes for almost every operation. The browser is created `async` and in the case of `WinForms` only when the controls `Hwnd` (handle) is created. (Remembering this discussion was originally about **WinForms**)

In `WPF` the browser is only created when it has a width/height. (CEF won't create a browser with size 0,0).

The `OffScreen` implementation is the only one that will attempt to create the browser when the `ChromiumWebBrowser` instance is created (again it's created async)

> ```
> <wpf:ChromiumWebBrowser x:Name="Browser" Address="..." Initialized="Wrapper_Browser_Initialized" />
> ```

The `WPF` `Initailzied` even is not the same as the browser being initialized. The `IsBrowserInitializedProperty` dependency property can be used to track the browser state (in addition to the event mentioned above).

There's examples of waiting for the browser to initialize in https://github.com/cefsharp/CefSharp/wiki/General-Usage#devtools

- `LoadingStateChanged` event is useful in a lot of cases
- `WaitForInitialLoadAsync` is useful in cases where you wish to ensure the browser has finished loading the first page.


> Plus the fact that the `Initialized` event is indeed being raised, at least as far as the CEF scaffold is concerned, we think it is initialized.

The `WPF` control has been initialized, different to `Chromium` being initialized. That's a standard `WPF` event (comes from `FrameworkElement`), it's not one that we can control.

Anyone can edit the `WIKI`, contributions are welcome and encouraged, if you'd like to see the documentation improved the please contribute!
